firewood/src/merkle/stream.rs

```diff
@@ -54,236 +48,301 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
-                let mut last_node = root_node;
-                let mut parents = vec![];
-                let leaf = loop {
-                    match last_node.inner() {
-                        NodeType::Branch(branch) => {
-                            let Some((leftmost_position, leftmost_address)) = branch
-                                .children
-                                .iter()
-                                .enumerate()
-                                .filter_map(|(i, addr)| addr.map(|addr| (i, addr)))
-                                .next()
-                            else {
-                                // we already exhausted the branch node. This happens with an empty trie
-                                // ... or a corrupt one
-                                return if parents.is_empty() {
-                                    // empty trie
-                                    Poll::Ready(None)
-                                } else {
-                                    // branch with NO children, not at the top
-                                    Poll::Ready(Some(Err(api::Error::InternalError(Box::new(
-                                        MerkleError::ParentLeafBranch,
-                                    )))))
-                                };
-                            };
-
-                            let next = self
-                                .merkle
-                                .get_node(leftmost_address)
-                                .map_err(|e| api::Error::InternalError(Box::new(e)))?;
-
-                            parents.push((last_node, leftmost_position as u8));
-
-                            last_node = next;
-                        }
-                        NodeType::Leaf(leaf) => break leaf,
-                        NodeType::Extension(_) => todo!(),
-                    }
-                };
 
-                // last_node should have a leaf; compute the key and value
-                let current_key = key_from_parents_and_leaf(&parents, leaf);
+                // always put the sentinal node in parents
+                let mut parents = vec![(root, 0)];
+
+                let next_result = find_next_result(merkle, &mut parents)
+                    .map_err(|e| api::Error::InternalError(Box::new(e)))
+                    .transpose();
 
-                self.key_state = IteratorState::Iterating { last_node, parents };
+                self.key_state = IteratorState::Iterating { parents };
 
-                current_key
+                Poll::Ready(next_result)
             }
+
             IteratorState::StartAtKey(key) => {
-                // TODO: support finding the next key after K
-                let root_node = self
-                    .merkle
-                    .get_node(self.merkle_root)
+                if key.is_empty() {
+                    self.key_state = IteratorState::StartAtBeginning;
```

Comment:
I'm not sure this is strictly necessary, but since there is code here, can you add a test case for it? I don't see one where we start an iterator with an empty key.

Reply:
deleted